nucliadb/src/nucliadb/common/cache.py

```diff
@@ -0,0 +1,171 @@
+# Copyright (C) 2021 Bosutech XXI S.L.
+#
+# nucliadb is offered under the AGPL v3.0 and as commercial software.
+# For commercial licensing, contact us at info@nuclia.com.
+#
+# AGPL:
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+
+import asyncio
+from abc import ABC, abstractmethod
+from contextvars import ContextVar
+from functools import cached_property
+from typing import Generic, Optional, TypeVar
+
+from lru import LRU
+
+from nucliadb.ingest.orm.resource import Resource as ResourceORM
+from nucliadb_protos.utils_pb2 import ExtractedText
+from nucliadb_telemetry.metrics import Gauge
+
+# specific metrics per cache type
+cached_resources = Gauge("nucliadb_cached_resources")
+cached_extracted_texts = Gauge("nucliadb_cached_extracted_texts")
+
+
+T = TypeVar("T")
+
+
+class Cache(Generic[T], ABC):
+    """Low-level bounded cache implementation with access to per-key async locks
+    in case cache users want to lock concurrent access.
+
+    This cache is measured using a mandatory metric all subclasses must define.
+
+    """
+
+    def __init__(self, cache_size: int) -> None:
+        self.cache: LRU[str, T] = LRU(cache_size, callback=self._evicted_callback)
+        self.locks: dict[str, asyncio.Lock] = {}
+
+    def get(self, key: str) -> Optional[T]:
+        return self.cache.get(key)
+
+    # Get a lock for a specific key. Locks will be evicted at the same time as
+    # key-value pairs
+    def get_lock(self, key: str) -> asyncio.Lock:
+        return self.locks.setdefault(key, asyncio.Lock())
+
+    def set(self, key: str, value: T):
+        len_before = len(self.cache)
+
+        self.cache[key] = value
+
+        len_after = len(self.cache)
+        if len_after - len_before > 0:
+            self._cache_size_metric.inc(len_after - len_before)
+
+    def contains(self, key: str) -> bool:
+        return key in self.cache
+
+    def clear(self):
+        self._cache_size_metric.dec(len(self.cache))
+        self.cache.clear()
+        self.locks.clear()
+
+    def __del__(self):
+        # we want to clear the cache before deleting the object and set the
+        # metric appropriately
+        self.clear()
+
+    @abstractmethod
+    @cached_property
+    def _cache_size_metric(self) -> Gauge: ...
+
+    def _evicted_callback(self, key: str, value: T):
```

Comment:
more than the gauges, I think i'd be more interesting to know when there are evictions.
Can you add a counter metric here?